**Data Merging and missing values treatment**

In [ ]:
#Import Necessary libraries
import pandas as pd
import numpy as np

In [ ]:
#Loading files
players_df = pd.read_excel("/content/players.xlsx")
games_df =pd.read_excel("/content/games.xlsx")
games_lineups_df =pd.read_excel("/content/game_lineups.xlsx")
games_events_df =pd.read_excel("/content/game_events.xlsx")
appearances_df =pd.read_excel("/content/appearances.xlsx")

In [ ]:
#Players_Data
players_df.head() #reading top 5 rows


player_id                 name  last_season  current_club_id  \
0     124732  John Anthony Brooks         2023              533   
1     223047      Emerson Hyndman         2018              903   
2     307781         Lynden Gooch         2016              289   
3     370846         Timothy Weah         2023              506   
4     484756   Djordje Mihailovic         2023             1090   

           player_code country_of_birth date_of_birth        sub_position  \
0  john-anthony-brooks          Germany    1993-01-28         Centre-Back   
1      emerson-hyndman    United States    1996-04-09    Central Midfield   
2         lynden-gooch    United States    1995-12-24        Right Winger   
3         timothy-weah    United States    2000-02-22      Right Midfield   
4   djordje-mihailovic    United States    1998-11-10  Attacking Midfield   

   position   foot  height_in_cm  market_value_in_eur  \
0  Defender   left         194.0            2000000.0   
1  Midfield  right         170.0             800000.0   
2    Attack  right         173.0            1000000.0   
3  Midfield  right         183.0           14000000.0   
4  Midfield    NaN         178.0            5000000.0   

   highest_market_value_in_eur contract_expiration_date           agent_name  
0                   15000000.0               2024-06-30                ROGON  
1                    1800000.0                      NaT           OmniSports  
2                    1000000.0               2024-06-30  Unique Sports Group  
3                   15000000.0               2028-06-30    BS Group - BS Law  
4                    6000000.0               2026-06-30  YMU Management Ltd.

In [ ]:
#Games_data
games_df.head() #reading top 5 rows


game_id competition_id  season         round       date  home_club_goals  \
0  2222734            RU1    2012  19. Matchday 2012-12-07                2   
1  2224572            DK1    2012   3. Matchday 2012-07-28                1   
2  2224628            DK1    2012  22. Matchday 2013-03-08                2   
3  2224655            DK1    2012  22. Matchday 2013-03-10                0   
4  2224729            DK1    2012   1. Matchday 2012-07-15                0   

   away_club_goals  home_club_position  away_club_position  \
0                3                11.0                10.0   
1                2                12.0                 1.0   
2                0                 3.0                10.0   
3                3                 6.0                12.0   
4                1                 9.0                 5.0   

  home_club_manager_name away_club_manager_name               stadium  \
0        Miodrag Bozovic       Slavoljub Muslin               Olimp-2   
1            Jess Thorup       Lars Sondergaard      Blue Water Arena   
2             Colin Todd       Lars Sondergaard  Cepheus Park Randers   
3           Peter Rensen        Auri Skarbalius            Ceres Park   
4        Auri Skarbalius            Troels Bech       Brondby Stadium   

   attendance              referee  home_club_formation  away_club_formation  \
0      5700.0     Sergey Kuznetsov                  NaN                  NaN   
1      5345.0     Michael Svendsen                  NaN                  NaN   
2      3481.0         Jakob Kehlet                  NaN                  NaN   
3      7967.0          Kenn Hansen                  NaN                  NaN   
4     10806.0  Lars Christoffersen                  NaN                  NaN   

              home_club_name   away_club_name aggregate competition_type  
0                  FK Rostov     FK Krasnodar  02:03:00  domestic_league  
1                 Esbjerg fB   Sonderjyske Fc  01:02:00  domestic_league  
2       Randers Fodbold Club   Sonderjyske Fc  02:00:00  domestic_league  
3  Aarhus Gymnastik Forening    BRA NDBYERNES  00:03:00  domestic_league  
4                 Brondby IF  Odense Boldklub  00:01:00  domestic_league

In [ ]:
#Games_lineups_data
games_lineups_df.head() #reading top 5 rows


game_lineups_id  game_id             type  number  \
0  f2570d1504fc02f4b6c7608e8dcf89a3  4087925      substitutes      34   
1  f5f0da93ea8e1d8bdd799658e7c8f7cb  4087928  starting_lineup      13   
2  31a4d12ec23d604779d909d26c1b5410  4087929      substitutes      26   
3  776dcbef98651450db76723cb7e3b4df  4087935      substitutes      26   
4  6a35ef7495303f29e7f85dbd54547fb1  4087936  starting_lineup      13   

   player_id     player_name  team_captain     position  
0     242284   Ethan Horvath             0   Goalkeeper  
1     145466        Tim Ream             0  Centre-Back  
2     578539  Chris Richards             0  Centre-Back  
3     578539  Chris Richards             0  Centre-Back  
4     145466        Tim Ream             1  Centre-Back

In [ ]:
#Games_events_data
games_events_df.head() #reading top 5 rows


game_event_id       date  game_id  minute  \
0  c6a3c088ed8a38d4ce074dd73b20d3da 2012-08-19  2221641      62   
1  02d605a5c2dc4f9a6721daa583fa5405 2012-08-26  2222536      54   
2  b56c2e2e087cddb3cfe9e3d340975df9 2012-11-18  2222707      79   
3  4a15d1fff4f476f48bb60092c61641d5 2012-11-23  2222721      72   
4  daa97877f7edf2fda885b411d7197921 2013-05-17  2222782      63   

            type  player_id  \
0  Substitutions       1335   
1          Cards       1321   
2  Substitutions     104203   
3  Substitutions     104203   
4          Goals     104203   

                                         description  player_in_id  \
0                                     , Not reported       45668.0   
1                             1. Yellow card  , Foul           NaN   
2                                         , Tactical       25740.0   
3                                         , Tactical       15583.0   
4  , Right-footed shot, 1. Goal of the Season Ass...           NaN   

   player_assist_id  
0               NaN  
1               NaN  
2               NaN  
3               NaN  
4           15077.0

In [ ]:
#Appearances_data
appearances_df.head() #reading top 5 rows


appearance_id  game_id  player_id       date       player_name  \
0  2224728_119169  2224728     119169 2012-07-13   Aron Johannsson   
1  2224732_161244  2224732     161244 2012-07-14     Conor O'Brien   
2   2224729_39467  2224729      39467 2012-07-15  Clarence Goodson   
3  2232104_119169  2232104     119169 2012-07-19   Aron Johannsson   
4   2219794_39475  2219794      39475 2012-07-22    Sacha Kljestan   

  competition_id  yellow_cards  red_cards  goals  assists  minutes_played  
0            DK1             0          0      0        0              90  
1            DK1             0          0      1        0              90  
2            DK1             0          0      0        0              90  
3            ELQ             0          0      0        0              74  
4           BESC             0          0      0        0               3

In [ ]:

#check columns names and shape of every data
print("Player data",players_df.columns,players_df.shape)
print("Games data",games_df.columns,games_df.shape)
print("Games_lineups data",games_lineups_df.columns,games_lineups_df.shape)
print("Games_events data",games_events_df.columns,games_events_df.shape)
print("Appearnces data",appearances_df.columns,appearances_df.shape)


Player data Index(['player_id', 'name', 'last_season', 'current_club_id', 'player_code',
       'country_of_birth', 'date_of_birth', 'sub_position', 'position', 'foot',
       'height_in_cm', 'market_value_in_eur', 'highest_market_value_in_eur',
       'contract_expiration_date', 'agent_name'],
      dtype='object') (152, 15)
Games data Index(['game_id', 'competition_id', 'season', 'round', 'date',
       'home_club_goals', 'away_club_goals', 'home_club_position',
       'away_club_position', 'home_club_manager_name',
       'away_club_manager_name', 'stadium', 'attendance', 'referee',
       'home_club_formation', 'away_club_formation', 'home_club_name',
       'away_club_name', 'aggregate', 'competition_type'],
      dtype='object') (3268, 20)
Games_lineups data Index(['game_lineups_id', 'game_id', 'type', 'number', 'player_id',
       'player_name', 'team_captain', 'position'],
      dtype='object') (219, 8)
Games_events data Index(['game_event_id', 'date', 'game_id', 'minute', 'typ

**Merging the data**

In [ ]:
merged_df1 = pd.merge(appearances_df, games_df, on='game_id', how='inner')

In [ ]:
merged_df2 = pd.merge(merged_df1, games_events_df, on='game_id', how='inner')

In [ ]:
merged_df3 = pd.merge(merged_df2,players_df,left_on='player_id_x',right_on='player_id',how='inner')

In [ ]:
merged_df4 = pd.merge(merged_df3,games_lineups_df,on="player_id",how="inner")

In [ ]:
#check the missing values
merged_df4.isnull().sum()

appearance_id                     0
game_id_x                         0
player_id_x                       0
date_x                            0
player_name_x                     0
competition_id_x                  0
yellow_cards                      0
red_cards                         0
goals                             0
assists                           0
minutes_played                    0
competition_id_y                  0
season                            0
round                             0
date_y                            0
home_club_goals                   0
away_club_goals                   0
home_club_position              718
away_club_position              718
home_club_manager_name            0
away_club_manager_name            0
stadium                           0
attendance                      396
referee                          32
home_club_formation            4199
away_club_formation            4199
home_club_name                  209
away_club_name                   22
aggregate                         0
competition_type                  0
game_event_id                     0
date                              0
minute                            0
type_x                            0
player_id_y                       0
description                    2658
player_in_id                   1446
player_assist_id               4167
player_id                         0
name                              0
last_season                       0
current_club_id                   0
player_code                       0
country_of_birth                  0
date_of_birth                     0
sub_position                      0
position_x                        0
foot                              0
height_in_cm                      0
market_value_in_eur               4
highest_market_value_in_eur       0
contract_expiration_date         12
agent_name                     2016
game_lineups_id                   0
game_id_y                         0
type_y                            0
number                            0
player_name_y                     0
team_captain                      0
position_y                        0
dtype: int64

In [ ]:
merged_df4.shape

(4199, 60)

In [ ]:
merged_df4.head()

appearance_id  game_id_x  player_id_x     date_x     player_name_x  \
0  2231426_49723    2231426        49723 2012-09-21  Timothy Chandler   
1  2231426_49723    2231426        49723 2012-09-21  Timothy Chandler   
2  2231426_49723    2231426        49723 2012-09-21  Timothy Chandler   
3  2231426_49723    2231426        49723 2012-09-21  Timothy Chandler   
4  2231426_49723    2231426        49723 2012-09-21  Timothy Chandler   

  competition_id_x  yellow_cards  red_cards  goals  assists  ...  \
0               L1             0          0      0        0  ...   
1               L1             0          0      0        0  ...   
2               L1             0          0      0        0  ...   
3               L1             0          0      0        0  ...   
4               L1             0          0      0        0  ...   

   highest_market_value_in_eur contract_expiration_date  agent_name  \
0                    5000000.0               2025-06-30         NaN   
1                    5000000.0               2025-06-30         NaN   
2                    5000000.0               2025-06-30         NaN   
3                    5000000.0               2025-06-30         NaN   
4                    5000000.0               2025-06-30         NaN   

                    game_lineups_id game_id_y       type_y  number  \
0  8d717b1017e72d89d2ee1112bb2d7466   4096025  substitutes      22   
1  d53eb3fd74523d6dd4536653c145ae96   4096042  substitutes      22   
2  3efc21c645e46b2f411bcdb1710aa444   4096061  substitutes      22   
3  cc79e78c570ce19872dfed125d324407   4162615  substitutes      22   
4  21de58902f547e231bf2604be9fa996f   4162639  substitutes      22   

      player_name_y  team_captain  position_y  
0  Timothy Chandler             0  Right-Back  
1  Timothy Chandler             0  Right-Back  
2  Timothy Chandler             0  Right-Back  
3  Timothy Chandler             0  Right-Back  
4  Timothy Chandler             0  Right-Back  

[5 rows x 60 columns]

In [ ]:
#missing value treatment
#Drop columns with more than 1000 missing values
columns_to_drop = merged_df4.columns[merged_df4.isnull().sum() > 1000]
merged_treated_df = merged_df4.drop(columns=columns_to_drop)

#Unneccessary columns
columns_to_drop =["competition_id_y","game_id_y",'player_id_y',"date_x","date_y","player_name_x"]
merged_treated_df=merged_treated_df.drop(columns_to_drop,axis=1)

#Treat remaining missing values
for column in merged_treated_df.columns:
    if merged_treated_df[column].isnull().sum() > 0:
        if  merged_treated_df[column].dtype == 'object':  # Categorical
           merged_treated_df[column].fillna(merged_treated_df[column].mode()[0], inplace=True)  # fill with mode
        else:  # Numerical
            merged_treated_df[column].fillna(merged_treated_df[column].mean(), inplace=True) #fill with mean

<ipython-input-21-b0d1449061eb>:16: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  merged_treated_df[column].fillna(merged_treated_df[column].mean(), inplace=True) #fill with mean
<ipython-input-21-b0d1449061eb>:14: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doi

In [ ]:
#after missing values treatment check the missing values
merged_treated_df.isnull().sum()

appearance_id                  0
game_id_x                      0
player_id_x                    0
competition_id_x               0
yellow_cards                   0
red_cards                      0
goals                          0
assists                        0
minutes_played                 0
season                         0
round                          0
home_club_goals                0
away_club_goals                0
home_club_position             0
away_club_position             0
home_club_manager_name         0
away_club_manager_name         0
stadium                        0
attendance                     0
referee                        0
home_club_name                 0
away_club_name                 0
aggregate                      0
competition_type               0
game_event_id                  0
date                           0
minute                         0
type_x                         0
player_id                      0
name                           0
last_season                    0
current_club_id                0
player_code                    0
country_of_birth               0
date_of_birth                  0
sub_position                   0
position_x                     0
foot                           0
height_in_cm                   0
market_value_in_eur            0
highest_market_value_in_eur    0
contract_expiration_date       0
game_lineups_id                0
type_y                         0
number                         0
player_name_y                  0
team_captain                   0
position_y                     0
dtype: int64

In [ ]:
#Final data Shape
merged_treated_df.shape

(4199, 48)

In [ ]:
#check duplicate values
duplicates = merged_treated_df.duplicated()
print(duplicates)

# If you want to see only the duplicate rows
duplicate_rows = merged_treated_df[merged_treated_df.duplicated()]
print(duplicate_rows)

0       False
1       False
2       False
3       False
4       False
        ...  
4194    False
4195    False
4196    False
4197    False
4198    False
Length: 4199, dtype: bool
Empty DataFrame
Columns: [appearance_id, game_id_x, player_id_x, competition_id_x, yellow_cards, red_cards, goals, assists, minutes_played, season, round, home_club_goals, away_club_goals, home_club_position, away_club_position, home_club_manager_name, away_club_manager_name, stadium, attendance, referee, home_club_name, away_club_name, aggregate, competition_type, game_event_id, date, minute, type_x, player_id, name, last_season, current_club_id, player_code, country_of_birth, date_of_birth, sub_position, position_x, foot, height_in_cm, market_value_in_eur, highest_market_value_in_eur, contract_expiration_date, game_lineups_id, type_y, number, player_name_y, team_captain, position_y]
Index: []

[0 rows x 48 columns]


In [ ]:
#covert the file into CSV
merged_treated_df.to_csv('Football_data_merged_cleaned_data.csv', index=False)

In [ ]:
Football_data=pd.read_csv("/content/Football_data_merged_cleaned_data.csv")
Football_data.head()

appearance_id  game_id_x  player_id_x competition_id_x  yellow_cards  \
0  2231426_49723    2231426        49723               L1             0   
1  2231426_49723    2231426        49723               L1             0   
2  2231426_49723    2231426        49723               L1             0   
3  2231426_49723    2231426        49723               L1             0   
4  2231426_49723    2231426        49723               L1             0   

   red_cards  goals  assists  minutes_played  season  ... height_in_cm  \
0          0      0        0              70    2012  ...        186.0   
1          0      0        0              70    2012  ...        186.0   
2          0      0        0              70    2012  ...        186.0   
3          0      0        0              70    2012  ...        186.0   
4          0      0        0              70    2012  ...        186.0   

   market_value_in_eur  highest_market_value_in_eur  \
0             700000.0                    5000000.0   
1             700000.0                    5000000.0   
2             700000.0                    5000000.0   
3             700000.0                    5000000.0   
4             700000.0                    5000000.0   

        contract_expiration_date                   game_lineups_id  \
0  2025-06-30 00:00:00.000000000  8d717b1017e72d89d2ee1112bb2d7466   
1  2025-06-30 00:00:00.000000000  d53eb3fd74523d6dd4536653c145ae96   
2  2025-06-30 00:00:00.000000000  3efc21c645e46b2f411bcdb1710aa444   
3  2025-06-30 00:00:00.000000000  cc79e78c570ce19872dfed125d324407   
4  2025-06-30 00:00:00.000000000  21de58902f547e231bf2604be9fa996f   

        type_y number     player_name_y  team_captain  position_y  
0  substitutes     22  Timothy Chandler             0  Right-Back  
1  substitutes     22  Timothy Chandler             0  Right-Back  
2  substitutes     22  Timothy Chandler             0  Right-Back  
3  substitutes     22  Timothy Chandler             0  Right-Back  
4  substitutes     22  Timothy Chandler             0  Right-Back  

[5 rows x 48 columns]

**Final Interpretation:**we merged 5 data files. we got whole data..then we perform missing values Treatment and drop some unnecessary columns.Finally we got 4199 rows and 48 columns